src/dtls13.c

```diff
@@ -0,0 +1,1968 @@
+            fragLength);
+
+        ret = Dtls13SendOneFragmentRtx(ssl, fsm, ssl->dtls13FragHandshakeType,
+            maxFragment, output, recordLength, 0);
+        if (ret == WANT_WRITE) {
+            ssl->dtls13FragOffset += fragLength;
+            return ret;
+        }
+
+        if (ret != 0) {
+            Dtls13FreeFragmentsBuffer(ssl);
+            return ret;
+        }
+
+        ssl->dtls13FragOffset += fragLength;
+        remainingSize -= fragLength;
+    }
+
+    /* we sent all fragments */
+    Dtls13FreeFragmentsBuffer(ssl);
+    return 0;
+}
+
+static int Dtls13SendFragmented(WOLFSSL *ssl, byte *message, word16 length,
+    enum HandShakeType handshake_type, int hash_output)
+{
+    int rlHeaderLength;
+    byte isEncrypted;
+    int messageSize;
+    int ret;
+
+    if (ssl->dtls13SendingFragments != 0)  {
+        WOLFSSL_MSG(
+            "dtls13_send_fragmented() invoked while already sending fragments");
+        return BAD_STATE_E;
+    }
+
+    isEncrypted = Dtls13TypeIsEncrypted(handshake_type);
+    rlHeaderLength = Dtls13GetRlHeaderLength(ssl, isEncrypted);
+
+    if (length < rlHeaderLength)
+        return INCOMPLETE_DATA;
+
+    /* DTLSv1.3 do not consider fragmentation for hash transcript. Build the
+       hash now pretending fragmentation will not happen */
+    if (hash_output) {
+        ret = HashRaw(ssl, message + rlHeaderLength, length - rlHeaderLength);
+        if (ret != 0)
+            return ret;
+    }
+
+    messageSize = length - rlHeaderLength - DTLS_HANDSHAKE_HEADER_SZ;
+
+    ssl->dtls13FragmentsBuffer.buffer =
+        (byte *)XMALLOC(messageSize, ssl->heap, DYNAMIC_TYPE_TMP_BUFFER);
+
+    if (ssl->dtls13FragmentsBuffer.buffer == NULL)
+        return MEMORY_E;
+
+    XMEMCPY(ssl->dtls13FragmentsBuffer.buffer,
+        message + rlHeaderLength + DTLS_HANDSHAKE_HEADER_SZ, messageSize);
+
+    ssl->dtls13MessageLength = messageSize;
+    ssl->dtls13FragHandshakeType = handshake_type;
+    ssl->dtls13SendingFragments = 1;
+
+    return Dtls13SendFragmentedInternal(ssl);
+}
+
+/**
+ * dtls13RlAddCiphertextHeader() - add record layer header in the buffer
+ * @ssl: ssl object
+ * @out: output buffer where to put the ehader
+ * @length: length of the record
+ */
+int Dtls13RlAddCiphertextHeader(WOLFSSL *ssl, byte *out, size_t length)
+{
+    Dtls13RecordCiphertextHeader *hdr;
+    word32 seqNumber;
+
+    if (out == NULL)
+        return BAD_FUNC_ARG;
+
+    if (ssl->dtls13EncryptEpoch == NULL)
+        return BAD_STATE_E;
+
+    hdr = (Dtls13RecordCiphertextHeader *)out;
+
+    hdr->unifiedHdrFlags = FIXED_BITS;
+    hdr->unifiedHdrFlags |= (ssl->dtls13EncryptEpoch->epochNumber & EE_MASK);
+
+    /* include 16-bit seq */
+    hdr->unifiedHdrFlags |= S_BIT;
+    /* include 16-bit length */
+    hdr->unifiedHdrFlags |= L_BIT;
+
+    seqNumber = ssl->dtls13EncryptEpoch->nextSeqNumberLo;
+    hdr->sequenceNumber[0] = (seqNumber >> 16) & 0xff;
+    hdr->sequenceNumber[1] = seqNumber & 0xff;
```

Comment:
I'm fairly certain I made a mistake here since it should be `>> 8`. That's why its best to use `c16toa`.